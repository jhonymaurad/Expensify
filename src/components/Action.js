import React from 'react';

const Action = props => {
  return (
    <div>
      <button
        onClick={props.handlePick}
        disabled={!props.hasOption}
        className="big-button"
      >
        What should I do?
      </button>
    </div>
  );
};

export default Action;
