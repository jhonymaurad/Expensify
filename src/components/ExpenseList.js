import React from 'react';
import ExpenseListItem from './ExpenseListItem';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = props => (
  <div>
    {props.expenses.length === 0 ? (
      <p>No expenses</p>
    ) : (
      props.expenses.map(expense => (
        <ExpenseListItem expense={expense} key={expense.id} />
      ))
    )}
  </div>
);

const mapStateToProps = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters) // Applied filters
  };
};

export default connect(mapStateToProps)(ExpenseList);
