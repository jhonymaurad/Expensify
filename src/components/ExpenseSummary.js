import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export function ExpenseSummary(props) {
  return (
    <div>
      {props.expenseCount > 1 ? (
        <p>
          Viewing {props.expenseCount} expenses totalling ${props.expenseTotal}
        </p>
      ) : (
        <p>
          Viewing {props.expenseCount} expense totalling ${props.expenseTotal}
        </p>
      )}
    </div>
  );
}

const mapStateToProps = state => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expenseTotal: selectExpensesTotal(visibleExpenses)
  };
};

export default connect(mapStateToProps)(ExpenseSummary);
