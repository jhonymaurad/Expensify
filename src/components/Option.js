import React from 'react';

const Option = ({ option, handleDeleteOption, count }) => {
  return (
    <div className="option">
      <p className="option__text">
        {count}. {option}
      </p>
      <button
        className="button button--link"
        onClick={e => {
          handleDeleteOption(option);
        }}
      >
        Remove
      </button>
    </div>
  );
};

export default Option;
