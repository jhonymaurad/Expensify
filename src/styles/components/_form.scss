.form {
  display: flex;
  flex-direction: column;
  > * {
    margin-bottom: $m-size;
  }
}
.form__error {
  margin: 0 0 $m-size 0;
  font-style: italic;
}

// import 'normalize.css/normalize.css';
// import './styles/styles.scss';
// import 'react-dates/initialize';
