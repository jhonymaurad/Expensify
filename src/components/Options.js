import React from 'react';
import Option from './Option';

const Options = props => {
  const { options } = props;
  return (
    <div>
      <div className="widget-header">
        <h3 className="widget-header__title">Your Options</h3>
        <button
          onClick={props.handleDeleteOptions}
          className="button button--link"
        >
          remove all
        </button>
      </div>

      {props.options.length === 0 && (
        <p className="widget-message">Please add an option to get started!</p>
      )}
      <ol>
        {options.map((option, index) => (
          <Option
            option={option}
            count={index + 1}
            key={index}
            handleDeleteOption={props.handleDeleteOption}
          />
        ))}
      </ol>
    </div>
  );
};

export default Options;
