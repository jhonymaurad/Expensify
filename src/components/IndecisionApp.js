import React from 'react';

import AddOption from './AddOption';
import Options from './Options';
import Header from './Header';
import Action from './Action';
import OptionModal from './OptionModal';

class IndecisionApp extends React.Component {
  state = {
    options: [],
    selectedOption: undefined
  };

  onClearOption = () => {
    this.setState(() => ({
      selectedOption: undefined
    }));
  };

  componentDidMount() {
    try {
      const json = localStorage.getItem('options');
      const options = JSON.parse(json);
      if (options) this.setState(() => ({ options }));
    } catch (error) {
      console.error(error);
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevState.options.length !== this.state.options.length) {
      const json = JSON.stringify(this.state.options);
      localStorage.setItem('options', json);
    }
  }

  handleDeleteOptions = () => {
    this.setState(() => ({
      options: []
    }));
  };

  handleDeleteOption = optionToRemove => {
    this.setState(prevState => ({
      options: prevState.options.filter(option => optionToRemove !== option)
    }));
  };

  handlePick = () => {
    const random = Math.floor(Math.random() * this.state.options.length);
    const selectedOption = this.state.options[random];
    this.setState(() => ({
      selectedOption
    }));
  };

  handleAddOption = option => {
    if (!option) {
      return 'Enter a valid option';
    } else if (this.state.options.indexOf(option) > -1) {
      return 'this option already exists';
    }

    this.setState(prevState => ({
      options: [...prevState.options, option] //prevState.options.concat(option)
    }));
  };

  render() {
    const subtitle = 'Put your life in the hands of a computer';
    const { options } = this.state;

    return (
      <div>
        <Header subtitle={subtitle} />
        <div className="container">
          <Action
            hasOption={this.state.options.length > 0}
            handlePick={this.handlePick}
          />
          <div className="widget">
            <Options
              options={options}
              handleDeleteOptions={this.handleDeleteOptions}
              handleDeleteOption={this.handleDeleteOption}
            />
            <AddOption handleAddOption={this.handleAddOption} />
          </div>
        </div>
        <OptionModal
          selectedOption={this.state.selectedOption}
          onClearOption={this.onClearOption}
        />
      </div>
    );
  }
}

IndecisionApp.defaultProps = {
  options: []
};

export default IndecisionApp;
