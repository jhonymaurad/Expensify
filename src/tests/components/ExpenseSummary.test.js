import React from 'react';
import { ExpenseSummary } from '../../components/ExpenseSummary';
import { shallow } from 'enzyme';
import { expenses } from '../fixtures/expenses';

test('should render ExpenseSummary correctly with one expense', () => {
  const wrapper = shallow(
    <ExpenseSummary expenseCount={1} expenseTotal={235} />
  );
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseSummary correctly with multiple expenses', () => {
  const wrapper = shallow(
    <ExpenseSummary expenseCount={expenses.length} expenseTotal={114195} />
  );
  expect(wrapper).toMatchSnapshot();
});
