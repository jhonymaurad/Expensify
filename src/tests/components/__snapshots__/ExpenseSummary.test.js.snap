// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary correctly with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    3
     
    expenses
     totalling 
    $1,141.95
  </h1>
</div>
`;

exports[`should render ExpenseSummary correctly with one expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $2.35
  </h1>
</div>
`;
